core/src/main/scala/org/http4s/blaze/pipeline/stages/TimeoutStageBase.scala

```diff
@@ -22,14 +24,21 @@ abstract class TimeoutStageBase[T](timeout: Duration, exec: TickWheelExecutor) e
 
   private val killswitch = new Runnable {
     override def run(): Unit = {
+      logger.debug(s"Timeout of $timeout triggered. Killing pipeline.")
       sendOutboundCommand(Disconnect)
       sendInboundCommand(Disconnected)
     }
   }
 
-  protected def setAndCancel(next: Cancellable): Unit = {
+  @tailrec
+  final def setAndCancel(next: Cancellable): Unit = {
     val prev = lastTimeout.getAndSet(next)
-    if (prev != null) prev.cancel()
+    if (prev == closedTag && next != closedTag) {
+      // woops... we submitted a new cancellation when we were closed!
```

Comment:
I don't have a whole lot to add, but every time I see a "woops" comment, I'm inclined to look for a log.warn so maybe the developer can see where they woopsed.